File: 207.cpp
/*
There are a total of n courses you have to take, labeled from 0 to n-1.

Some courses may have prerequisites, for example to take course 0 you have to
first take course 1, which is expressed as a pair: [0,1]

Given the total number of courses and a list of prerequisite pairs, is it
possible for you to finish all courses?

Example 1:

Input: 2, [[1,0]]
Output: true
Explanation: There are a total of 2 courses to take.
             To take course 1 you should have finished course 0. So it is
possible. Example 2:

Input: 2, [[1,0],[0,1]]
Output: false
Explanation: There are a total of 2 courses to take.
             To take course 1 you should have finished course 0, and to take
course 0 you should also have finished course 1. So it is impossible. Note:

The input prerequisites is a graph represented by a list of edges, not adjacency
matrices. Read more about how a graph is represented. You may assume that there
are no duplicate edges in the input prerequisites.
*/

#include <iostream>
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<int> degree(numCourses, 0);
        for (auto&& p : prerequisites) {
            degree[p.first]++;
        }

        queue<int> q;
        for (int i = 0; i < degree.size(); ++i) {
            if (degree[i] == 0) {
                q.push(i);
            }
        }

        int cnt = 0;
        while (!q.empty()) {
            int prere = q.front();
            q.pop();
            cnt++;
            for (auto& p : prerequisites) {
                if (p.second == prere) {
                    p.second = -1;
                    if (--degree[p.first] == 0) {
                        q.push(p.first);
                    }
                }
            }
        }

        return cnt == numCourses;
    }
};

int main() {
    Solution s;
    vector<pair<int, int>> pre = {{1, 0}};
    int numCourses = 2;
    cout << s.canFinish(numCourses, pre) << endl;

    return 0;
}

File: Python/207.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        q = deque()
        degrees = [0]*numCourses
        for p in prerequisites:
            degrees[p[0]] +=1
        for i in range(numCourses):
            if degrees[i] == 0:
                q.append(i)
        cnt = 0
        while q:
            prer = q.popleft()
            cnt +=1
            for p in prerequisites:
                if p[1] == prer:
                    p[1] = -1
                    degrees[p[0]] -=1
                    if degrees[p[0]] == 0:
                        q.append(p[0])
        return cnt == numCourses



class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = defaultdict(set)
        todo = {i:set() for i in range(numCourses)}
        for pre in prerequisites:
            graph[pre[1]].add(pre[0])
            todo[pre[0]].add(pre[1])

        q = deque()
        for t in todo:
            if todo[t] == set():
                q.append(t)

        while q:
            pre = q.popleft()
            for c in graph[pre]:
                todo[c].discard(pre)
                if todo[c] == set():
                    q.append(c)
            todo.pop(pre)
        return todo == {}
